### PR TITLE
Feature/general defaults

### PR DESCRIPTION
**Pull request by roesslerj**
Please finish PR#50 before, as this code bases on that branch...

### CODE REVIEW

File: src/main/java/de/retest/web/DefaultValuesProvider.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		if ( attributeValue == null || attributeValue.trim().isEmpty() ) {
+			return true;
+		}
+		if ( attributeValue.equals( "auto" ) ) {
+			return true;
+		}
+		if ( attributeValue.equals( "normal" ) ) {
+			return true;
+		}
+		if ( attributeValue.equals( "0px" ) ) {
```
Comment (retest-sonar-cloud) [inline src/main/java/de/retest/web/DefaultValuesProvider.java]: Code Smell Code Smell: Replace this if-then-else statement by a single return statement. ([squid:S1126](https://sonarcloud.io/organizations/retest/rules#rule_key=squid%3AS1126))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=de.retest%3Arecheck-web&pullRequest=52&issues=AWURSaL-yxJLWg29qApz&open=AWURSaL-yxJLWg29qApz)

[ISSUE_KEY]: AWURSaL-yxJLWg29qApz

File: src/main/java/de/retest/web/DefaultValuesProvider.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	public DefaultValuesProvider() {
+		try ( final InputStream url = getClass().getResourceAsStream( DEFAULTS_FILE_PATH ) ) {
+			defaultValues = readAttributesConfigFromFile( url );
+		} catch ( final Exception e ) {
+			throw new ReadDefaultValuesException( e );
+		}
+	}
+
+	private Map<String, Map<String, String>> readAttributesConfigFromFile( final InputStream in ) throws IOException {
+		final Map<String, Map<String, String>> defaultValues = new HashMap<>();
```
Comment (retest-sonar-cloud) [inline src/main/java/de/retest/web/DefaultValuesProvider.java]: Code Smell Code Smell: Rename "defaultValues" which hides the field declared at line 32. ([squid:HiddenFieldCheck](https://sonarcloud.io/organizations/retest/rules#rule_key=squid%3AHiddenFieldCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=de.retest%3Arecheck-web&pullRequest=52&issues=AWURSaL-yxJLWg29qAp0&open=AWURSaL-yxJLWg29qAp0)

[ISSUE_KEY]: AWURSaL-yxJLWg29qAp0

File: src/main/java/de/retest/web/DefaultValuesProvider.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	public static final class ReadDefaultValuesException extends RuntimeException {
+
+		private static final long serialVersionUID = 1L;
+
+		public ReadDefaultValuesException( final Exception e ) {
+			super( "Could not load default CSS values from " + DEFAULTS_FILE_PATH, e );
+		}
+	}
+
+	public static final String DEFAULTS_FILE_PATH = "/defaults.yaml";
```
Comment (retest-sonar-cloud) [inline src/main/java/de/retest/web/DefaultValuesProvider.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/retest/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=de.retest%3Arecheck-web&pullRequest=52&issues=AWURSaL-yxJLWg29qApy&open=AWURSaL-yxJLWg29qApy)

[ISSUE_KEY]: AWURSaL-yxJLWg29qApy

File: src/main/resources/defaults.yaml [hunk 4]
```diff
@@ -1,5 +1,38 @@
 a:link:
   - text-decoration: underline
+all:
```
Comment (diba1013) [inline src/main/resources/defaults.yaml]: We should at least try to keep it alphabetical...

File: src/main/resources/defaults.yaml [hunk 5]
```diff
@@ -1,5 +1,38 @@
...
+  - margin-left: 0px
+  - overflow: visible
+  - padding-top: 0px
+  - padding-bottom: 0px
+  - padding-left: 0px
+  - padding-right: 0px
+  - text-align: start
+  - visibility: visible
+
   - cursor: auto
```
Comment (diba1013) [inline src/main/resources/defaults.yaml]: Did you move the cursor outside of `a:link`?

File: src/main/java/de/retest/web/WebElementPeer.java [hunk 6]
```diff
@@ -53,7 +53,10 @@ protected IdentifyingAttributes retrieveIdentifyingAttributes() {
 		attributes.add( new PathAttribute( Path.fromString( path ) ) );
 		attributes.add( new SuffixAttribute( path.substring( path.lastIndexOf( '[' ) + 1, path.lastIndexOf( ']' ) ) ) );
 		attributes.add( new StringAttribute( "type", webData.getAsString( "tagName" ) ) );
-		attributes.add( new TextAttribute( "text", webData.getAsString( "text" ) ) );
+		final String text = webData.getAsString( "text" );
+		if ( text != null && !text.trim().isEmpty() ) {
```
Comment (diba1013) [inline src/main/java/de/retest/web/WebElementPeer.java]: Is there a `StringUtils.isEmpty()`?

File: src/main/resources/attributes.yaml [hunk 7]
```diff
@@ -1,6 +1,13 @@
...
   - animation
-  - background
+  - background-color
+  - background-image
+  - background-position
+  - background-size
+  - background-repeat
+  - background-origin
+  - background-clip
+  - background-attachment
```
Comment (diba1013) [inline src/main/resources/attributes.yaml]: Please keep it alphabetical

File: src/main/resources/attributes.yaml [hunk 8]
```diff
@@ -1,17 +1,24 @@
...
+  - animation-name
+  - animation-duration
+  - animation-timing-function
+  - animation-delay
+  - animation-iteration-count
+  - animation-direction
+  - animation-fill-mode
+  - animation-play-state
+  - background-attachment
+  - background-clip
```
Comment (diba1013) [inline src/main/resources/attributes.yaml]: ABC.. 😄

File: src/main/java/de/retest/web/DefaultValuesProvider.java [hunk 9]
```diff
@@ -83,9 +83,6 @@ public boolean isDefault( final String tag, final String attribute, final String
 		if ( attributeValue.equals( "auto" ) ) {
 			return true;
 		}
-		if ( attributeValue.equals( "none" ) ) {
```
Comment (diba1013) [inline src/main/java/de/retest/web/DefaultValuesProvider.java]: https://www.w3schools.com/cssref/css_default_values.asp there are `none` or is it a placeholder?

Comment (roesslerj) [reply src/main/java/de/retest/web/DefaultValuesProvider.java]: Hab ich ja eingetragen...

Comment (roesslerj) [reply src/main/java/de/retest/web/DefaultValuesProvider.java]: in der defaults.yaml.

**Pull request merged**